### PR TITLE
#22325 validation of org. unit termination

### PR DESCRIPTION
**Pull request by danchr**
This adds support for rejecting invalid termination requests, along with a few other, minor fixes.

### CODE REVIEW

File: mora/service/orgunit.py [hunk 1]
```diff
@@ -660,9 +662,85 @@ def terminate_org_unit(unitid):
...
+                children['total'],
+            )
+        else:
+            msg = 'cannot terminate unit with {} active roles'.format(
+                len(roles),
+            )
+
+        raise common.ValidationException(
+            msg,
+            child_units=children['items'],
```
[the remark this answers is not in the document]
Comment (danchr) [reply mora/service/orgunit.py]: > Does the UI need to know about this? Do they care? The issue is clearly stated in the message, so this seems excessive.

The original UI mentioned the child units by name — we're essentially giving the UI the information they need to present the same error message.

File: mora/service/common.py [hunk 2]
```diff
@@ -38,7 +39,29 @@
...
-class FieldTypes(enum.Enum):
+class ValidationException(werkzeug.exceptions.Conflict):
+    def __init__(self, message: typing.Optional[str]=None, **context):
+        super().__init__(message)
+
+        self.message = message or 'Validation failed'
+        self.context = context
+
+    def to_json(self):
+        return {
```
[the remark this answers is not in the document]
Comment (danchr) [reply mora/service/common.py]: > What is this structure? Is this a new common structure for messages to the UI?

It is mostly similar to what we report in `handle_invalid_usage()` in `app.py` — but also intended as a new, generic infrastructure for validation, yes.

[the remark this answers is not in the document]
Comment (cmoesgaard) [reply mora/service/common.py]: That makes sense. Is `success` ever going to be anything other than `False`?

[the remark this answers is not in the document]
Comment (cmoesgaard) [reply mora/service/common.py]: And ditto `cause`.

File: mora/service/orgunit.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (cmoesgaard) [reply mora/service/orgunit.py]: Ah, okay!

File: mora/service/common.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (danchr) [reply mora/service/common.py]: > That makes sense. Is `success` ever going to be anything other than `False`?

Not at the moment — and coming to think of it, the frontend would have to do `r.success === false` to check it, which is kind of horrible. We should probably replace it with an `error` field with a value of true, and add that to `handle_invalid_usage` as well.

> And ditto cause.

No, it's never anything other than `"validation"`, given that we don't have any other well-defined exceptions currently. We should probably improve our error handling in general, though — the frontenders typically find errors somewhat unhelpful.

[the remark this answers is not in the document]
Comment (cmoesgaard) [reply mora/service/common.py]: Okay. Given that this is supposed to be a general message format, and that it is supposed to similar to the one in `app.py` I suggest we extract the creation of these, to a factory of sorts.

File: mora/exceptions.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+
+class ValidationError(BaseError):
+    description = 'Invalid input.'
+    cause = 'validation'
+    code = 400
+
+
+class NotFoundError(BaseError):
+    description = 'No found.'
```
Comment (cmoesgaard) [inline mora/exceptions.py]: _Not_ found

**Pull request merged**